File: app/src/main/java/org/helllabs/android/xmp/player/viewer/InstrumentViewer.java
package org.helllabs.android.xmp.player.viewer;

import org.helllabs.android.xmp.R;
import org.helllabs.android.xmp.service.ModInterface;
import org.helllabs.android.xmp.util.Log;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.os.RemoteException;

public class InstrumentViewer extends Viewer {
    private static final String TAG = "InstrumentViewer";
    private final Paint[] insPaint, barPaint;
    private final int fontSize, fontHeight, fontWidth;
    private String[] insName;
    private final Rect rect = new Rect();

    public InstrumentViewer(final Context context) {
        super(context);

        fontSize = getResources().getDimensionPixelSize(R.dimen.instrumentview_font_size);

        insPaint = new Paint[8];
        for (int i = 0; i < 8; i++) {
            final int val = 120 + 10 * i;
            insPaint[i] = new Paint();
            insPaint[i].setARGB(255, val, val, val);
            insPaint[i].setTypeface(Typeface.MONOSPACE);
            insPaint[i].setTextSize(fontSize);
            insPaint[i].setAntiAlias(true);
        }

        barPaint = new Paint[8];
        for (int i = 0; i < 8; i++) {
            final int val = 15 * i;
            barPaint[i] = new Paint();
            barPaint[i].setARGB(255, val / 4, val / 2, val);
        }

        fontWidth = (int) insPaint[0].measureText("X");
        fontHeight = fontSize * 14 / 10;
    }

    @Override
    public void setup(final ModInterface modPlayer, final int[] modVars) {
        super.setup(modPlayer, modVars);

        final int insNum = modVars[4];

        try {
            insName = modPlayer.getInstruments();
        } catch (RemoteException e) {
            Log.e(TAG, "Can't get instrument name");
        }

        setMaxY(insNum * fontHeight + fontHeight / 2);
    }

    @Override
    public void update(final Info info, final boolean paused) {
        super.update(info, paused);

        Canvas canvas = null;

        try {
            canvas = surfaceHolder.lockCanvas(null);
            if (canvas != null) {
                synchronized (surfaceHolder) {
                    doDraw(canvas, modPlayer, info);
                }
            }
        } finally {
            // do this in a finally so that if an exception is thrown
            // during the above, we don't leave the Surface in an
            // inconsistent state
            if (canvas != null) {
                surfaceHolder.unlockCanvasAndPost(canvas);
            }
        }
    }

    private void doDraw(final Canvas canvas, final ModInterface modPlayer, final Info info) {
        final int chn = modVars[3];
        final int ins = modVars[4];

        // Clear screen
        canvas.drawColor(Color.BLACK);

        for (int i = 0; i < ins; i++) {
            final int y = (i + 1) * fontHeight - (int) posY;
            final int width = (canvasWidth - 3 * fontWidth) / chn;
            int maxVol;

            // Don't draw if not visible
            if (y < 0 || y > canvasHeight + fontHeight) {
                continue;
            }

            maxVol = 0;
            for (int j = 0; j < chn; j++) {

                if (isMuted[j]) {
                    continue;
                }

                if (info.instruments[j] == i) {
                    final int x = 3 * fontWidth + width * j;
                    int vol = info.volumes[j] / 8;
                    if (vol > 7) {
                        vol = 7;
                    }
                    rect.set(x, y - fontSize + 1, x + width * 8 / 10, y + 1);
                    canvas.drawRect(rect, barPaint[vol]);
                    if (vol > maxVol) {
                        maxVol = vol;
                    }
                }
            }

            if (insName != null) {
                canvas.drawText(insName[i], 0, y, insPaint[maxVol]);
            }
        }
    }
}

File: app/src/main/java/org/helllabs/android/xmp/player/viewer/InstrumentViewer.kt
package org.helllabs.android.xmp.player.viewer

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.os.RemoteException
import androidx.core.graphics.ColorUtils
import org.helllabs.android.xmp.R
import org.helllabs.android.xmp.Xmp
import org.helllabs.android.xmp.util.color
import org.helllabs.android.xmp.util.logD
import org.helllabs.android.xmp.util.logE

@SuppressLint("ViewConstructor")
class InstrumentViewer(context: Context, val background: Int) : Viewer(context, background) {

    private val startBlue: Int = resources.color(R.color.accent)

    private lateinit var insName: Array<String>
    private val barPaint = arrayListOf<Paint>()
    private val fontHeight: Int
    private val fontSize: Int = resources.getDimensionPixelSize(R.dimen.instrumentview_font_size)
    private val fontWidth: Int
    private val insPaint = arrayListOf<Paint>()
    private val rect = Rect()

    // Draw Loop Variables
    private var chn: Int = 0
    private var drawWidth: Int = 0
    private var drawX: Int = 0
    private var drawY: Int = 0
    private var ins: Int = 0
    private var maxVol: Int = 0
    private var vol: Int = 0

    init {
        // White text volume shades
        for (i in 0..10) {
            val value: Float = (i / 10f)
            logD("Text Value $i: $value")
            insPaint.add(
                Paint().apply {
                    color = ColorUtils.blendARGB(Color.GRAY, Color.WHITE, value)
                    alpha = 255
                    typeface = Typeface.MONOSPACE
                    textSize = fontSize.toFloat()
                    isAntiAlias = true
                }
            )
        }

        // Blue bar volume shades
        for (i in 10 downTo 0) {
            val value: Float = (i / 10f)
            logD("Bar Value $i: $value")
            barPaint.add(
                Paint().apply {
                    color = ColorUtils.blendARGB(startBlue, background, value)
                    alpha = 255
                }
            )
        }

        fontWidth = insPaint[0].measureText("X").toInt()
        fontHeight = fontSize * 14 / 10
    }

    override fun setup(modVars: IntArray) {
        super.setup(modVars)
        logD("Viewer Setup")

        try {
            insName = Xmp.getInstruments()
        } catch (e: RemoteException) {
            logE("Can't get instrument name")
        }

        setMaxY(modVars[4] * fontHeight + fontHeight / 2)
    }

    override fun update(info: Info?, paused: Boolean) {
        super.update(info, paused)

        requestCanvasLock { canvas ->
            doDraw(canvas, info)
        }
    }

    @Suppress("UNUSED_PARAMETER")
    private fun doDraw(canvas: Canvas, info: Info?) {
        chn = modVars[3]
        ins = modVars[4]

        // Clear screen
        canvas.drawColor(bgColor)

        for (i in 0 until ins) {
            drawY = (i + 1) * fontHeight - posY.toInt()
            drawWidth = (canvasWidth - 3 * fontWidth) / chn
            maxVol = 0

            // Don't draw if not visible
            if (drawY < 0 || drawY > canvasHeight + fontHeight) {
                continue
            }

            for (j in 0 until chn) {
                if (isMuted[j]) {
                    continue
                }
                if (info!!.instruments[j] == i) {
                    drawX = 3 * fontWidth + drawWidth * j
                    vol = info.volumes[j] / 6

                    // Clamp
                    if (vol > 60) {
                        vol = 60
                    }

                    rect.set(drawX, drawY - fontSize + 4, drawX + drawWidth * 8 / 10, drawY + 8)
                    canvas.drawRect(rect, barPaint[vol])
                    if (vol > maxVol) {
                        maxVol = vol
                    }
                }
            }
            canvas.drawText(insName[i], 0f, drawY.toFloat(), insPaint[maxVol])
        }
    }
}
